

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tutorial: Assemble, bin and asses eukaryotic bins &mdash; EukCC  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Frequently asked questions:" href="../FAQ.html" />
    <link rel="prev" title="EukCC" href="../index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> EukCC
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#binning">Binning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optional-filtering-for-euks">Optional: Filtering for euks</a></li>
<li class="toctree-l2"><a class="reference internal" href="#estimate-bin-completeness-and-contamination-using-eukcc">Estimate bin completeness and contamination using EukCC</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#eukcc-ouput">EukCC ouput</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../FAQ.html">FAQ</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">EukCC</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Tutorial: Assemble, bin and asses eukaryotic bins</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/usage/tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tutorial-assemble-bin-and-asses-eukaryotic-bins">
<h1>Tutorial: Assemble, bin and asses eukaryotic bins<a class="headerlink" href="#tutorial-assemble-bin-and-asses-eukaryotic-bins" title="Permalink to this headline">¶</a></h1>
<p>We assume for this tutorial, that you have sequenced
and assembled your metagenomic reads without prior filtering
using your assembler of choice (metaSPAdes, flye, megahit, etc.).</p>
<div class="section" id="binning">
<h2>Binning<a class="headerlink" href="#binning" title="Permalink to this headline">¶</a></h2>
<p>To bin eukaryotic metagenomic sequences you need to use a binner with
no prior towards prokariotic sequences. Thus you will not be able
to use neither maxbin2 not metabat. We suggest the use of CONCOCT, but
any unbiased binner will work.</p>
<p>Follow the github page of CONCOCT if you want:</p>
<p><a class="reference external" href="https://github.com/BinPro/CONCOCT">https://github.com/BinPro/CONCOCT</a></p>
<p>We assume for this tutorial that your bins are in the folder
fasta_bins
and end with .fa.</p>
</div>
<div class="section" id="optional-filtering-for-euks">
<h2>Optional: Filtering for euks<a class="headerlink" href="#optional-filtering-for-euks" title="Permalink to this headline">¶</a></h2>
<p>After binning you may want to reduce the number of bins by looking for bins
with large amounts of eukaryotic DNA. This step will bias your analysis
towards what will be detected by EukRep as such, but can be helpful
if you deal with a large number of bins.</p>
<p>You could for example filter for bins with a minimum number of
eukaryotic bases and/or a ratio of eukaryotic to unclassified or bacterial
DNA. We noticed a ratio of at least 20 % eukaryotic DNA will already make sure
to exclude any bacterial bins.</p>
<p>This can be done using our helper script “filter_euk_bins.py”</p>
<p>Make sure you have EukRep (<a class="reference external" href="https://github.com/patrickwest/EukRep">https://github.com/patrickwest/EukRep</a>) installed.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>filter_euk_bins.py fasta_bins/*.fa
</pre></div>
</div>
<p>This will create a file <cite>assignment.csv</cite> containing comma seperated
columns with stats about the bin including one column ‘eukaryotic’, which
can help you decide.</p>
<p>The assignent file looks like:</p>
<table border="1" class="docutils" id="id1">
<caption><span class="caption-text">assignment.csv</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
<col width="14%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">path</th>
<th class="head">binname</th>
<th class="head">eukaryotic</th>
<th class="head">eukbases</th>
<th class="head">bacbases</th>
<th class="head">unassigned</th>
<th class="head">sum</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>fasta_bins/0.fa</td>
<td>0.fa</td>
<td>True</td>
<td>2593713</td>
<td>0</td>
<td>0</td>
<td>2593713</td>
</tr>
<tr class="row-odd"><td>fasta_bins/1.fa</td>
<td>1.fa</td>
<td>False</td>
<td>36682</td>
<td>394058</td>
<td>3639</td>
<td>434379</td>
</tr>
</tbody>
</table>
<p>Try adjusting the ratios using the flags, if you are unhappy with the defaults.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>filter_euk_bins.py fasta_bins/*.fa --eukratio <span class="m">0</span>.5 --bacratio <span class="m">0</span>.4
</pre></div>
</div>
<p>All options are described in the help:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>filter_euk_bins.py -h
</pre></div>
</div>
</div>
<div class="section" id="estimate-bin-completeness-and-contamination-using-eukcc">
<h2>Estimate bin completeness and contamination using EukCC<a class="headerlink" href="#estimate-bin-completeness-and-contamination-using-eukcc" title="Permalink to this headline">¶</a></h2>
<p>Now we can launch EukCC for each eukaryotic bin. We suggest submitting
EukCC to a compute cluster, in this example we use LSF bsub:</p>
<p>We assume that you located the eukccdb in your home folder, you might need
to adjust that path.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="k">while</span> <span class="nv">IFS</span><span class="o">=</span>, <span class="nb">read</span> -r bin col2 euk col4
<span class="k">do</span>
    <span class="k">if</span> <span class="o">[</span> <span class="nv">$euk</span> <span class="o">==</span> <span class="s2">&quot;True&quot;</span> <span class="o">]</span><span class="p">;</span> <span class="k">then</span>
        <span class="nv">NAME</span><span class="o">=</span><span class="k">$(</span>basename <span class="nv">$bin</span><span class="k">)</span>
        bsub -M <span class="m">80000</span> -J eukcc_<span class="nv">$bin</span> -n <span class="m">16</span> <span class="se">\</span>
            <span class="s2">&quot;eukcc --db </span><span class="nv">$HOME</span><span class="s2">/eukccdb --ncores 16 \</span>
<span class="s2">             --plot \</span>
<span class="s2">             --ncorespplacer 1 --outdir eukcc/</span><span class="nv">$NAME</span><span class="s2"> </span><span class="nv">$bin</span><span class="s2">&quot;</span>
    <span class="k">fi</span>
<span class="k">done</span> &lt; asignment.csv
</pre></div>
</div>
<p>In this example we allocate 80 Gb in memory for EukCC and use 16 cores for
gene prediction and annotation. To reduce memory consumption we only use a
single thread for pplacer.</p>
<div class="section" id="eukcc-ouput">
<h3>EukCC ouput<a class="headerlink" href="#eukcc-ouput" title="Permalink to this headline">¶</a></h3>
<p>EukCC will create a structure like this:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ tree -L <span class="m">2</span> eukcc/1.fa
    eukcc/1.fa
    ├── eukcc.tsv
    └── workfiles
        ├── gmes
        ├── hmmer
        └── pplacer
</pre></div>
</div>
<p>The main output is the file <cite>eukcc.tsv</cite>. It will contain predictions with up
to three sets chosen to best encompass the phylogenetic location of the bin.</p>
<table border="1" class="docutils" id="id2">
<caption><span class="caption-text">eukcc.csv</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
<col width="9%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">completeness</th>
<th class="head">contamination</th>
<th class="head">node</th>
<th class="head">n</th>
<th class="head">ngenomes</th>
<th class="head">cover</th>
<th class="head">nPlacements</th>
<th class="head">taxid</th>
<th class="head">lineage</th>
<th class="head">taxidlineage</th>
<th class="head">file</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>91.2</td>
<td>0.0</td>
<td>node218</td>
<td>523</td>
<td>56</td>
<td>15</td>
<td>17</td>
<td>4892</td>
<td>root_cellular.organisms_Eukaryota_Opisthokonta_Fungi_Dikarya_Ascomycota_saccharomyceta_Saccharomycotina_Saccharomycetes_Saccharomycetales</td>
<td>1_131567_2759_33154_4751_451864_4890_716545_147537_4891_4892</td>
<td>GCF_003054445.1-95-fakeGenome-0</td>
</tr>
<tr class="row-odd"><td>91.96</td>
<td>0.0</td>
<td>node187</td>
<td>448</td>
<td>59</td>
<td>15</td>
<td>17</td>
<td>4892</td>
<td>root_cellular.organisms_Eukaryota_Opisthokonta_Fungi_Dikarya_Ascomycota_saccharomyceta_Saccharomycotina_Saccharomycetes_Saccharomycetales</td>
<td>1_131567_2759_33154_4751_451864_4890_716545_147537_4891_4892</td>
<td>GCF_003054445.1-95-fakeGenome-0</td>
</tr>
<tr class="row-even"><td>92.42</td>
<td>0.17</td>
<td>node161</td>
<td>594</td>
<td>372</td>
<td>15</td>
<td>17</td>
<td>4892</td>
<td>root_cellular.organisms_Eukaryota_Opisthokonta_Fungi_Dikarya_Ascomycota_saccharomyceta_Saccharomycotina_Saccharomycetes_Saccharomycetales</td>
<td>1_131567_2759_33154_4751_451864_4890_716545_147537_4891_4892</td>
<td>GCF_003054445.1-95-fakeGenome-0</td>
</tr>
</tbody>
</table>
<p>In this table up to three quality estimates are given. These are the three
best sets found to estimate the quality of this MAG. The best set is the
first row. It can be interesting to look at more than one set as sometimes
a lower set gives a more robust estimate.</p>
<p>Most columns of the table will be self explanatory. We want to highlight a few
critical ones:</p>
<ul class="simple">
<li><strong>n</strong>: This is the number of protein-profiles used to estimate completeness
and contamination</li>
<li><strong>node</strong>: This is the location in the reference tree and can be used to
see if different mags are located in the same area of the tree.</li>
<li><strong>ngenomes</strong>: This is the number of reference genomes used to construct the
set used to estimate bin quality. A low number will suggest a less stable
estimate.</li>
<li><strong>nPlacements and cover</strong>: <cite>nPlacements</cite> is the number of proteins placed
in the reference tree and <cite>cover</cite> is the number of these placed below the
<cite>node</cite>-set used to estimate completeness.</li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../FAQ.html" class="btn btn-neutral float-right" title="Frequently asked questions:" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../index.html" class="btn btn-neutral float-left" title="EukCC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Paul Saary

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>